### ISSUE TITLE
Compile error with Nvidia HPC sdk

### ISSUE THREAD

**Issue by JudithYi** [ExtremeFLOW/neko#569]
How to build with nvidia hpc sdk? I met the problem to find cuda.  I used the following configure commend: 
```
../configure --with-cuda=/path/to/nvhpcsdk/21.11/Linux_x86_64/22.5/cuda/11.5 NVCC=/path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/cuda/11.5/bin/nvcc
```
or 
```
../configure --with-cuda=/path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/cuda/11.5 NVCC=/path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/cuda/11.5/bin/nvcc CUDA_LDFLAGS=/path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/cuda/11.5/lib64
```
The error from configure looks like: 
```
checking for nvcc... /path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/cuda/11.5/bin/nvcc
checking for cudaFree in -lcudart... no
configure: error: CUDA not found
```

But the libcudart.so is located in `/path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/cuda/11.5/lib64/`

And the following commend: 
```
nm -D /path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/cuda/11.5/lib64/libcudart.so | grep cudaFree
```

gives the output as:
```
000000000004e3c0 T cudaFree
000000000004e700 T cudaFreeArray
000000000006b0d0 T cudaFreeAsync
00000000000551b0 T cudaFreeAsync_ptsz
000000000004e560 T cudaFreeHost
000000000004e8a0 T cudaFreeMipmappedArray
```

**Comment by njansson** [comment]
Hmm, the flags given to `configure` are the correct ones.

Maybe `config.log` can shed some more light on the linking issue. The following should give the command that failed (and the error message)
```shell
grep -4 cudaFree config.log
```

**Comment by JudithYi** [comment]
Thank you for the hint.
In the `config.log` it turns out the problem is the Blas/Lapack libraries:
```
configure:8865: checking for cudaFree in -lcudart
configure:8890: /u/system/soft/SLE_15/packages/x86_64/nvhpcsdk/21.11/Linux_x86_64/21.11/cuda/11.5/bin/nvcc -o conftest -g -O2   -L/u/system/soft/SLE_15/packages/x86_64/nvhpcsdk/21.11/Linux_x86_64/21.11/cuda/11.5/lib64  conftest.c -lcudart   -llapack -lblas    >&5
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: cannot find -llapack
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: cannot find -lblas
collect2: error: ld returned 1 exit status
configure:8890: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "neko"
| #define PACKAGE_TARNAME "neko"
| #define PACKAGE_VERSION "0.4.99"
| #define PACKAGE_STRING "neko 0.4.99"
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| #define PACKAGE "neko"
| #define VERSION "0.4.99"
| #define HAVE_DERIVED_TYPE_IO 1
| #define HAVE_MPI_F08 1
| #define HAVE_MPI_PARAM_DTYPE 1
| #define HAVE_BLAS 1
| #define HAVE_LAPACK 1
| /* end confdefs.h.  */
|
| /* Override any GCC internal prototype to avoid an error.
|    Use char because int might match the return type of a GCC
|    builtin and then its argument prototype would still apply.  */
| #ifdef __cplusplus
| extern "C"
| #endif
| char cudaFree ();
| int
| main ()
| {
| return cudaFree ();
|   ;
|   return 0;
| }
configure:8899: result: no
configure:8915: error: CUDA not found
```
And with: 
```
../configure --with-cuda=/path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/cuda/11.5:/path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/compilers NVCC=/path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/cuda/11.5/bin/nvcc --with-blas=/path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/compilers/lib/libblas.so --with-lapack=/path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/compilers/lib/liblapack.so 
```
I got: 
```
configure:8865: checking for cudaFree in -lcudart
configure:8890: /path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/cuda/11.5/bin/nvcc -o conftest -g -O2   conftest.c -lcudart   /path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/compilers/lib/liblapack.so /path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/compilers/lib/libblas.so    >&5
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/compilers/lib/liblapack.so: undefined reference to `pgf90_str_cpy1'
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/compilers/lib/liblapack.so: undefined reference to `pgf90_strcmp_klen'
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/compilers/lib/liblapack.so: undefined reference to `pgf90_str_copy_klen'
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/compilers/lib/liblapack.so: undefined reference to `pghpf_maxloc_i8'
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/compilers/lib/liblapack.so: undefined reference to `pgf90_set_intrin_type_i8'
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /path/to/nvhpcsdk/21.11/Linux_x86_64/21.11/compilers/lib/liblapack.so: undefined reference to `__get_size_of'
collect2: error: ld returned 1 exit status
configure:8890: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "neko"
| #define PACKAGE_TARNAME "neko"
| #define PACKAGE_VERSION "0.4.99"
| #define PACKAGE_STRING "neko 0.4.99"
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| #define PACKAGE "neko"
| #define VERSION "0.4.99"
| #define HAVE_DERIVED_TYPE_IO 1
| #define HAVE_MPI_F08 1
| #define HAVE_MPI_PARAM_DTYPE 1
| #define HAVE_BLAS 1
| #define HAVE_LAPACK 1
| /* end confdefs.h.  */
|
| /* Override any GCC internal prototype to avoid an error.
|    Use char because int might match the return type of a GCC
|    builtin and then its argument prototype would still apply.  */
| #ifdef __cplusplus
| extern "C"
| #endif
| char cudaFree ();
| int
| main ()
| {
| return cudaFree ();
|   ;
|   return 0;
| }
configure:8899: result: no
configure:8915: error: CUDA not found
```
So for the Blas/Lapack, do we need it on GPU in case we wanna run the code on the GPUs? Could we use the cuBlas?

**Comment by JudithYi** [comment]
I am able to compile NEKO with gcc/10, openmpi/4, cuda/11.4, openmpi/4, and mkl for BLAS/LAPACK. The problem could be in the FORTRAN related library in nvhpcsdk/21.

**Comment by njansson** [comment]
Great! There was an issue in the configure script that could potentially try to link against Fortran libraries with nvcc. This was recently fixed in fixed in #575.

About the cuBlas question, blas/lapack is only used to solve a small system ( < 10 x 10) to obtain BDF coefficients in each time step, so this should be fine to have on the CPU.

**Comment by njansson** [comment]
An update, compiling Neko with NVIDIA HPC toolkit works with the latest toolkit release 22.11 and a minor patch that can be found under patches (see #723)
